src/cir.rs

```diff
@@ -0,0 +1,439 @@
+                            &mut operator_stack,
+                            &mut operand_stack,
+                        );
+                    }
+                }
+            },
+            LirInstruction::Predicate(p) => {
+                operand_stack.push(CirOperand::Predicate(p.clone()));
+                reduce_translation_stack(
+                    &mut cir.instructions,
+                    &mut operator_stack,
+                    &mut operand_stack,
+                );
+
+                index += 1;
+            }
+        }
+
+        if index >= lir.instructions.len() {
+            // end of LirProgram
+            break;
+        }
+    }
+    debug_assert_eq!(operator_stack.len(), 0);
+}
+
+fn execute_helper(
+    cir_instructions: &[CirInstruction],
+    index: usize,
+    ctx: &mut Context,
+    m: &mut Match,
+) -> bool {
+    match &cir_instructions[index] {
+        CirInstruction::AndIns(and) => {
+            let left_val = if is_index(&and.left) {
+                execute_helper(cir_instructions, and.left.as_index().unwrap(), ctx, m)
+            } else {
+                and.left.as_predicate().unwrap().execute(ctx, m)
+            };
+
+            if !left_val {
+                // short circuit
+                return false;
+            } else {
+                if is_index(&and.right) {
+                    return execute_helper(cir_instructions, and.right.as_index().unwrap(), ctx, m);
+                } else {
+                    return and.right.as_predicate().unwrap().execute(ctx, m);
+                }
+            }
+        }
+        CirInstruction::OrIns(or) => {
+            let left_val = if is_index(&or.left) {
+                execute_helper(cir_instructions, or.left.as_index().unwrap(), ctx, m)
+            } else {
+                or.left.as_predicate().unwrap().execute(ctx, m)
+            };
+
+            if left_val {
+                // short circuit
+                return true;
+            } else {
+                if is_index(&or.right) {
+                    return execute_helper(cir_instructions, or.right.as_index().unwrap(), ctx, m);
+                } else {
+                    return or.right.as_predicate().unwrap().execute(ctx, m);
+                }
+            }
+        }
+        CirInstruction::NotIns(not) => {
+            let right_val = if is_index(&not.right) {
+                execute_helper(cir_instructions, not.right.as_index().unwrap(), ctx, m)
+            } else {
+                not.right.as_predicate().unwrap().execute(ctx, m)
+            };
+            return !right_val;
+        }
+    }
+}
+
+impl Execute for CirProgram {
+    fn execute(&self, ctx: &mut Context, m: &mut Match) -> bool {
+        execute_helper(&self.instructions, self.instructions.len() - 1, ctx, m)
+    }
+}
+
+fn add_to_counter_helper(
+    cir_instructions: &[CirInstruction],
+    index: usize,
+    map: &mut HashMap<String, usize>,
+) {
+    match &cir_instructions[index] {
```

Comment:
Is that correct that we don't do iteration for `cir_instructions` here? It seems like that if we have a `AndIns(index, index)` in the tail of `cir_instructions`, than nothing will be done for the fields counter?